.gitlab-ci.yml

```
@@ -104,74 +151,145 @@ build:mender-client-acceptance-testing:
+        -f mender-client-acceptance-testing/Dockerfile
+        --push
+        mender-client-acceptance-testing
+    - docker save ${CI_REGISTRY_IMAGE}:${CONTAINER_TAG}-${CI_PIPELINE_ID} > qaTestingImage.tar
     # Upload to temporary S3 bucket
     - mender_ci_save_tmp_artifact qaTestingImage.tar
   artifacts:
     paths:
       - checksums
 
+build:aws-k8s-pipeline-toolbox:
+  tags:
+    - mender-qa-worker-generic-light
+  stage: build
+  image: docker
+  needs: []
+  variables:
+    CONTAINER_TAG: "aws-k8s-v1"
+    CI_KUBECTL_VERSION: "1.24.0"
+    CI_HELM_VERSION: "v3.9.0"
+    CI_KUBECONFORM_VERSION: "0.4.14"
+  services:
+    - docker:dind
+  script:
+    - echo "INFO - Building and Pushing ${CONTAINER_TAG}-${CI_PIPELINE_ID} to the registry ${CI_REGISTRY_IMAGE}"
+    - echo $CI_REGISTRY_PASSWORD | docker login -u $CI_REGISTRY_USER $CI_REGISTRY --password-stdin
+    - docker build 
+        --cache-from ${CI_REGISTRY_IMAGE}:${CONTAINER_TAG}
+        -t ${CI_REGISTRY_IMAGE}:${CONTAINER_TAG}-${CI_PIPELINE_ID}
+        --build-arg KUBECTL_VERSION=${CI_KUBECTL_VERSION}
+        --build-arg HELM_VERSION=${CI_HELM_VERSION}
+        --build-arg KUBECONFORM_VERSION=${CI_KUBECONFORM_VERSION}
+        -f aws-k8s-toolbox/Dockerfile
+        --push
+        aws-k8s-toolbox
+
 .template:publish:
   stage: publish
   services:
     - docker:dind
-  only:
-    - master
+  rules:
+    - if: $CI_COMMIT_BRANCH == "master"
   tags:
     - docker
   before_script:
     - echo -n $DOCKER_HUB_PASSWORD | docker login -u $DOCKER_HUB_USERNAME --password-stdin
+    - echo $CI_REGISTRY_PASSWORD | docker login -u $CI_REGISTRY_USER $CI_REGISTRY --password-stdin
 
 publish:raspbian_latest:
   extends: .template:publish
   dependencies:
     - build:raspbian_latest
+  variables:
+    CONTAINER_TAG: "rasbian_latest"
   script:
+    # backward compatibility: pushing to the Docker Hub
     - echo "publishing image to Docker Hub"
     - docker load -i raspbianImage.tar
-    - docker push $DOCKER_REPOSITORY:raspbian_latest
+    - docker tag ${CI_REGISTRY_IMAGE}:${CONTAINER_TAG}-${CI_PIPELINE_ID} $DOCKER_REPOSITORY:$CONTAINER_TAG
+    - docker push $DOCKER_REPOSITORY:$CONTAINER_TAG
+    # Gitlab registry
+    - *tag_n_push_to_gitlab_registry
 
 publish:acceptance-testing:
   extends: .template:publish
   dependencies:
     - build:acceptance-testing
+  variables:
+    CONTAINER_TAG: "acceptance-testing"
   script:
+    # backward compatibility: pushing to the Docker Hub
     - echo "publishing image to Docker Hub"
     - docker load -i acceptanceTestingImage.tar
-    - docker push $DOCKER_REPOSITORY:acceptance-testing
+    - docker tag ${CI_REGISTRY_IMAGE}:${CONTAINER_TAG}-${CI_PIPELINE_ID} $DOCKER_REPOSITORY:$CONTAINER_TAG
+    - docker push $DOCKER_REPOSITORY:$CONTAINER_TAG
+    # Gitlab registry
+    - *tag_n_push_to_gitlab_registry
 
 publish:backend-integration-testing:
   extends: .template:publish
   dependencies:
```

```suggestion
  extends: .template:publish
  variables:
    CONTAINER_TAG: "backend-integration-testing"  
  dependencies:
```
sorry I missed this earlier 🙈